File: 프로그래머스/2/87946. 피로도/피로도.java
import java.util.*;

class Solution {
    public boolean[] visited = new boolean[8];
    public int[] dMax = new int[100000];
    public int m = 0;
    public int[][] dungeons;
    
    public int solution(int k, int[][] dungeons) {
        this.dungeons = dungeons;
        
        backtracking(0, k);
        
        int max = 0;
        
        for(int i = 0; dMax[i] != 0; i++) {
            if(dMax[i] > max) max = dMax[i];
        }
        
        return max;
    }
    
    public void backtracking(int count, int k) {
        for(int i = 0; i < dungeons.length; i++) {
            if(k == 0) {
                dMax[m++] = count;
                return;
            }
            
            if(!visited[i] && k >= dungeons[i][0]) {
                visited[i] = true;
                backtracking(count + 1, k - dungeons[i][1]);
                visited[i] = false;
            }
            if(i == dungeons.length - 1) dMax[m++] = count;
        }
    }
}

File: 프로그래머스/2/87946. 피로도/피로도.py
def back_tracking(k, dungeons, visited, answer):
    
    for idx, dungeon in enumerate(dungeons):
        if any(idx == v for v in visited) or k < dungeon[0]:
            continue
        if k >= dungeon[0]:
            visited.append(idx)
            k -= dungeon[1]
            answer = max(answer, len(visited))
            
            answer = back_tracking(k, dungeons, visited, answer)
            
            visited.pop()
            k += dungeon[1]

    return answer

def solution(k, dungeons):
    answer = 0
    visited = []
    
    return back_tracking(k, dungeons, visited, answer)    
